cpp/serve/grammar/grammar_builder.h

```diff
@@ -0,0 +1,190 @@
+    grammar_->rule_expr_indptr_.push_back(grammar_->rule_expr_data_.size());
+    grammar_->rule_expr_data_.push_back(static_cast<int32_t>(rule_expr.kind));
+    grammar_->rule_expr_data_.insert(grammar_->rule_expr_data_.end(), rule_expr.data,
+                                     rule_expr.data + rule_expr.data_len);
+    return grammar_->rule_expr_indptr_.size() - 1;
+  }
+
+  /*!
+   * \brief One element of a character range, containing a lower and a upper bound. Both bounds are
+   * inclusive.
+   */
+  struct CharacterRangeElement {
+    int32_t lower;
+    int32_t upper;
+  };
+
+  /*! \brief Insert a RuleExpr for character range.*/
+  int32_t InsertCharacterRange(const std::vector<CharacterRangeElement>& elements) {
+    std::vector<int32_t> data;
+    for (const auto& range : elements) {
+      data.push_back(range.lower);
+      data.push_back(range.upper);
+    }
+    return InsertRuleExpr({DataKind::kCharacterRange, data.data(), data.size()});
+  }
+
+  /*! \brief Insert a RuleExpr for character range negation.*/
+  int32_t InsertNegCharacterRange(const std::vector<CharacterRangeElement>& elements) {
+    std::vector<int32_t> data;
+    for (const auto& range : elements) {
+      data.push_back(range.lower);
+      data.push_back(range.upper);
+    }
+    return InsertRuleExpr({DataKind::kNegCharacterRange, data.data(), data.size()});
+  }
+
+  /*! \brief Insert a RuleExpr for empty string.*/
+  int32_t InsertEmptyStr() { return InsertRuleExpr({DataKind::kEmptyStr, nullptr, 0}); }
+
+  /*! \brief Insert a RuleExpr for rule reference.*/
+  int32_t InsertRuleRef(int32_t rule_id) {
+    std::vector<int32_t> data;
+    data.push_back(rule_id);
+    return InsertRuleExpr({DataKind::kRuleRef, data.data(), data.size()});
+  }
+
+  /*! \brief Insert a RuleExpr for RuleExpr sequence.*/
+  int32_t InsertSequence(const std::vector<int32_t>& elements) {
+    std::vector<int32_t> data;
+    data.insert(data.end(), elements.begin(), elements.end());
+    return InsertRuleExpr({DataKind::kSequence, data.data(), data.size()});
+  }
+
+  /*! \brief Insert a RuleExpr for RuleExpr choices.*/
+  int32_t InsertChoices(const std::vector<int32_t>& choices) {
+    std::vector<int32_t> data;
+    data.insert(data.end(), choices.begin(), choices.end());
+    return InsertRuleExpr({DataKind::kChoices, data.data(), data.size()});
+  }
+
+  /*! \brief Get the rule_expr with the given id. */
+  RuleExpr GetRuleExpr(int32_t rule_expr_id) { return grammar_->GetRuleExpr(rule_expr_id); }
+
+  /****************** Rule handling ******************/
+
+  /*! \brief Insert a rule and return the rule id. */
+  int32_t InsertRule(const Rule& rule) {
+    int32_t id = grammar_->rules_.size();
+    auto rules = grammar_->rules_;
+    grammar_->rules_.push_back(rule);
+    rule_name_to_id_[rule.name] = id;
+    return id;
+  }
+
+  /*! \brief Get the rule with the given id. */
+  Rule& GetRule(int32_t rule_id) { return grammar_->rules_[rule_id]; }
```

Comment:
For grammar builder I just want to support the case where the rule need to be updated. This could be used by GrammarSimplifier. As an alternative we can `UpdateRule(rule_id, body)`, but I'm not sure that would be better

Reply:
We can return pointer instead of reference then. Or reuse the logic of `SetRuleBody`.